File: src/shortcuts_mcp/executor.py
from __future__ import annotations

import asyncio
import json
import subprocess
import time
from urllib.parse import quote

from .types import JsonValue


def _stringify_input(value: JsonValue) -> str:
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    return json.dumps(value)


def _applescript_literal(value: str) -> str:
    return json.dumps(value)


def _open_url(url: str, timeout: int | None) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        ["open", url],
        capture_output=True,
        text=True,
        timeout=timeout,
        check=False,
    )


async def run_via_applescript(
    name: str, input_value: JsonValue | None = None
) -> tuple[str, int, int]:
    start = time.perf_counter()
    name_literal = _applescript_literal(name)
    script = [
        'tell application "Shortcuts Events"',
    ]
    if input_value is None:
        script.append(f"    run the shortcut named {name_literal}")
    else:
        input_literal = _applescript_literal(_stringify_input(input_value))
        script.append(
            f"    run the shortcut named {name_literal} with input {input_literal}"
        )
    script.append("end tell")

    process = await asyncio.create_subprocess_exec(
        "osascript",
        "-e",
        "\n".join(script),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await process.communicate()
    output = stdout.decode().strip()
    stderr_text = stderr.decode().strip() if stderr else ""
    if stderr_text:
        output = f"{output}\n{stderr_text}".strip()
    elapsed_ms = int((time.perf_counter() - start) * 1000)
    returncode = process.returncode if process.returncode is not None else 1
    return output, elapsed_ms, returncode


async def run_via_url_scheme(
    name: str, input_value: JsonValue | None = None, timeout: int | None = None
) -> None:
    url = f"shortcuts://run-shortcut?name={quote(name)}"
    if input_value is not None:
        url += f"&input={quote(_stringify_input(input_value))}"

    loop = asyncio.get_running_loop()
    completed = await loop.run_in_executor(None, _open_url, url, timeout)
    if completed.returncode != 0:
        stderr = (completed.stderr or "").strip()
        stdout = (completed.stdout or "").strip()
        message = stderr or stdout or f"open returned {completed.returncode}"
        raise RuntimeError(message)

File: src/executor.ts
import type { JsonValue } from "./types.js";

export type CommandResult = {
  stdout: string;
  stderr: string;
  exitCode: number;
};

export type CommandOptions = {
  timeoutMs?: number | null;
};

export type CommandRunner = (
  command: string,
  args: string[],
  options?: CommandOptions,
) => Promise<CommandResult>;

export type RunResult = {
  output: string;
  elapsedMs: number;
  returncode: number;
};

const readStream = async (stream: ReadableStream<Uint8Array> | null): Promise<string> => {
  if (!stream) {
    return "";
  }
  const response = new Response(stream);
  return response.text();
};

const runCommand: CommandRunner = async (command, args, options = {}): Promise<CommandResult> => {
  const subprocess = Bun.spawn([command, ...args], {
    stdout: "pipe",
    stderr: "pipe",
  });

  const timeoutMs = options.timeoutMs ?? null;
  let timeoutId: ReturnType<typeof setTimeout> | null = null;
  let timedOut = false;

  if (timeoutMs !== null && timeoutMs > 0) {
    timeoutId = setTimeout(() => {
      timedOut = true;
      subprocess.kill();
    }, timeoutMs);
  }

  const [stdout, stderr, exitCode] = await Promise.all([
    readStream(subprocess.stdout),
    readStream(subprocess.stderr),
    subprocess.exited,
  ]);

  if (timeoutId) {
    clearTimeout(timeoutId);
  }

  if (timedOut) {
    throw new Error(`Command timed out after ${timeoutMs ?? 0} ms`);
  }

  return { stdout, stderr, exitCode };
};

export const stringifyInput = (value: JsonValue | null): string => {
  if (value === null) {
    return "";
  }
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value);
};

export const applescriptLiteral = (value: string): string => {
  return JSON.stringify(value);
};

export const buildApplescript = (name: string, inputValue: JsonValue | null): string => {
  const nameLiteral = applescriptLiteral(name);
  const script = ['tell application "Shortcuts Events"'];
  if (inputValue === null) {
    script.push(`    run the shortcut named ${nameLiteral}`);
  } else {
    const inputLiteral = applescriptLiteral(stringifyInput(inputValue));
    script.push(`    run the shortcut named ${nameLiteral} with input ${inputLiteral}`);
  }
  script.push("end tell");
  return script.join("\n");
};

export const buildShortcutUrl = (name: string, inputValue: JsonValue | null): string => {
  let url = `shortcuts://run-shortcut?name=${encodeURIComponent(name)}`;
  if (inputValue !== null) {
    url += `&input=${encodeURIComponent(stringifyInput(inputValue))}`;
  }
  return url;
};

export const runViaApplescript = async (
  name: string,
  inputValue: JsonValue | null = null,
  timeoutSeconds: number | null = null,
  runner: CommandRunner = runCommand,
): Promise<RunResult> => {
  const start = performance.now();
  const script = buildApplescript(name, inputValue);
  const timeoutMs = timeoutSeconds === null ? null : timeoutSeconds * 1000;
  const { stdout, stderr, exitCode } = await runner("osascript", ["-e", script], { timeoutMs });
  const trimmedStdout = stdout.trim();
  const trimmedStderr = stderr.trim();
  let output = trimmedStdout;
  if (trimmedStderr.length > 0) {
    output = `${output}\n${trimmedStderr}`.trim();
  }
  const elapsedMs = Math.round(performance.now() - start);
  return { output, elapsedMs, returncode: exitCode };
};

export const runViaUrlScheme = async (
  name: string,
  inputValue: JsonValue | null = null,
  timeoutSeconds: number | null = null,
  runner: CommandRunner = runCommand,
): Promise<void> => {
  const url = buildShortcutUrl(name, inputValue);
  const timeoutMs = timeoutSeconds === null ? null : timeoutSeconds * 1000;
  const { stdout, stderr, exitCode } = await runner("open", [url], {
    timeoutMs,
  });

  if (exitCode !== 0) {
    const trimmedStderr = stderr.trim();
    const trimmedStdout = stdout.trim();
    const message = trimmedStderr || trimmedStdout || `open returned ${exitCode}`;
    throw new Error(message);
  }
};
